synapse/events/spamcheck.py

```diff
@@ -30,19 +40,120 @@
 logger = logging.getLogger(__name__)
 
 
+def load_legacy_spam_checkers(hs: "synapse.server.HomeServer"):
+    """Wrapper that loads spam checkers configured using the old configuration, and
+    registers the spam checker hooks they implement.
+    """
+    spam_checkers = []  # type: List[Any]
+    api = hs.get_module_api()
+    for module, config in hs.config.spam_checkers:
+        # Older spam checkers don't accept the `api` argument, so we
+        # try and detect support.
+        spam_args = inspect.getfullargspec(module)
+        if "api" in spam_args.args:
+            spam_checkers.append(module(config=config, api=api))
+        else:
+            spam_checkers.append(module(config=config))
+
+    # The known spam checker hooks. If a spam checker module implements a method
+    # which name appears in this set, we'll want to register it.
+    spam_checker_methods = {
+        "check_event_for_spam",
+        "user_may_invite",
+        "user_may_create_room",
+        "user_may_create_room_alias",
+        "user_may_publish_room",
+        "check_username_for_spam",
+        "check_registration_for_spam",
+        "check_media_file_for_spam",
+    }
+
+    for spam_checker in spam_checkers:
+        # Get the properties (attributes and methods) of the module and filter it
+        # down to the supported method names.
+        props = dir(spam_checker)
+        supported_hooks = list(spam_checker_methods.intersection(props))
+
+        # Register the hooks through the module API.
+        hooks = {hook: getattr(spam_checker, hook) for hook in supported_hooks}
+
+        api.register_spam_checker_callbacks(**hooks)
+
+
 class SpamChecker:
     def __init__(self, hs: "synapse.server.HomeServer"):
-        self.spam_checkers = []  # type: List[Any]
-        api = hs.get_module_api()
-
-        for module, config in hs.config.spam_checkers:
-            # Older spam checkers don't accept the `api` argument, so we
-            # try and detect support.
-            spam_args = inspect.getfullargspec(module)
-            if "api" in spam_args.args:
-                self.spam_checkers.append(module(config=config, api=api))
-            else:
-                self.spam_checkers.append(module(config=config))
+        self._callbacks: Dict[str, List[Callable]] = {}
```

Comment:
I would honestly go with a separate list for each callback. I realise it'll make this longer, but I don't think it'll be too bad, and it'll mean you can replace `self._callbacks.get("user_may_create_room", [])` with `self._user_may_create_room_callbacks`.